### ISSUE TITLE
Filter a list

### ISSUE THREAD

**Issue by joelanman** [alphagov/govuk-design-system-backlog#133]
Also known as: faceted search

## What

Help users filter a list of items for to see things that are relevant to their task. For example filtering search results.

## Why

Services that use this pattern:

 - [GOV.UK search](https://www.gov.uk/search?q=tax)
 - GOV.UK Pay Admin
 - GOV.UK Notify Admin

## Anything else

- [Discussion on Design Wiki](https://paper.dropbox.com/doc/Search-filters-VxP46U2qgjrYnNwqXakKJ)

**Comment by ignaciaorellana** [comment]
There is a conversation in the designer mailing list that is talking about examples of filter or advance search components (relevant to this issue). There is an example from [Digital Market Place](https://www.digitalmarketplace.service.gov.uk/g-cloud/search?q=strategy+digital+cloud+as+a+service+) and [GOV.UK pages](https://www.gov.uk/drug-device-alerts)

**Comment by nickcolley** [comment]
You can see this component in GOV.UK Publishing at https://finder-frontend.herokuapp.com/component-guide/option-select

**Comment by adamsilver** [comment]
Here's some work we've done for HMCTS on filters:
https://github.com/hmcts/design-system-backlog/issues/43

**Comment by mikeash82** [comment]
@fordie commented on this issue on 26 Jun

This pattern allows a user to search or filter a list of items that have been returned by the system, for instance looking for only pubs or properties in a certain postcode that are in a list of properties. This is similar to search, but the user will already have been presented with a list of results that they want to reduce.

![find-a-client](https://user-images.githubusercontent.com/494985/41900478-bd3bc7ce-7926-11e8-9546-3387fc72dd1a.png)

![property filter](https://user-images.githubusercontent.com/494985/41901805-c6ad8574-7929-11e8-9c20-8b32e7b5e388.png)

@benwakefield commented on 27 Sep 

@stevenaproctor We're using this approach to refine results where usual closed checkbox filtering isn't appropriate, the prototype should be out for testing in the coming weeks. This UI is for an internal HMRC system fyi.

<img width="945" alt="screen shot 2018-09-27 at 07 58 21" src="https://user-images.githubusercontent.com/7381460/46128418-4ca40500-c22b-11e8-907a-11c56d918fee.png">

**Comment by timpaul** [comment]
We've moved this card to the 'To do' column of the community backlog because it's a dependency on the [Search for something](https://github.com/alphagov/govuk-design-system-backlog/issues/88) pattern, which is already in the 'To do' column.

**Comment by FarhadUX1** [comment]
> Faceted search tags (aka Chips, Tags, Pills) on GOV.UK finders e.g. https://www.gov.uk/service-standard-reports?keywords=MOT&assessment_date%5Bfrom%5D=&assessment_date%5Bto%5D=
> 
> ![screen shot 2019-01-21 at 14 54 47](https://user-images.githubusercontent.com/23356842/51482541-97bfc600-1d8e-11e9-9674-f4f14f72c9be.png)

Curious as to why the filters selected (in a rectangle) would be separated from the filters itself.  Happy to help with this design, as we are using filters for an activity feed.

**Comment by edwardhorsford** [comment]
It would be super if the collapsing filter component could be adopted - we've got a number of things you can filter by, and this helps keep the list manageable.

**Comment by joelanman** [comment]
**This has not yet been through any user research at all**

Here's a mockup of ideas for filtering GOV.UK search. I thought the accordion might work better with arrows here as people could understand plus and minus as adding filter (which is kind of true) and removing a filter (which is not true, you'd just be collapsing the group)

![image](https://user-images.githubusercontent.com/1132904/55155436-41388480-514f-11e9-8546-b00dfaed5961.png)

**Comment by campbellssource** [comment]
@joelanman I noticed that a previous implementation of search on gov.uk/search had a count next to each filter, Hinting at the consequence of selecting/deselecting the filter. I notice this isn't in the mock up or current version. have you had research suggest it isn't helpful?

**Comment by joelanman** [comment]
@campbellssource sorry for the long delay! There was no deliberate change there - I think at the point I made the screenshot we just hadn't implemented that aspect yet.

**Comment by adamsilver** [comment]
Here's the MOJ's filter a list pattern:
https://moj-design-system.herokuapp.com/patterns/filter-a-list

Big screens:
![image](https://user-images.githubusercontent.com/37163/69860718-bb954780-128e-11ea-873a-139e0c7792d6.png)

Small screens:
![image](https://user-images.githubusercontent.com/37163/69860761-cea81780-128e-11ea-93e2-0549f8768f1a.png)

**Comment by jonathaninch** [comment]
I'm looking into a set of filters that can be applied to a case load. I want to show how many cases there are for each filter (let's imagine there's 9 cases and there's 4 filters: case open, case closed, UK cases, international cases).
e.g.
- case open (6 cases)
- case closed (3 cases)
- UK cases (7 cases)
- international cases (2 cases)

What I want to do is change the case number in brackets if another filter affects that total - so in the above example if I select UK cases (of which there's, say, 7), then I'd get:
- case open (5 cases)
- case closed (2 cases)

In other words, because I've not selected international cases, those two have been removed from the totals within case open and case closed, and the totals in brackets have dynamically updated when the UK cases filter has been clicked. Hope that makes sense....

Has anyone done this or knows how to do it?

**Comment by kelliedesigner** [comment]
The Filters documentation has been updated in the MOJ Pattern Library:
https://design-patterns.service.justice.gov.uk/components/filter/

Feedback welcome!

**Comment by nedpowell** [comment]
For the CFS Service we updated the filters because the old filters had: 

- 'Clear filters' at the bottom which was hard for the user to see and reach 
- Low contrast between the text and background colours 
- No way to see what filters had been selected
- No way for the user to show/hide all sections at once 
- Outdated Accordions 

<img width="386" alt="Current filters" src="https://user-images.githubusercontent.com/92091904/165285058-a0347bb1-b611-4145-a100-2825e35294f8.png">

For the updated filters we:

- Based them off the new [Accordions](https://design-system.service.gov.uk/components/accordion/) 
- Took inspiration from the [MOJ's filter a list 'Selected filters'](https://design-patterns.service.justice.gov.uk/components/filter/) but decided the 'Apply filters' button wasn't necessary as 'Apply filters' could be done automatically, also the title 'Filters' doesn't add anything for the user. 
- Made 'Filter by' on each filter visually hidden for screen readers and added 'Search for' text under the funding stream and period to differ from the main 'Search' at the top of the screen. 
- A 'Clear search' button link was added so the user can clear the search separately from clearing the filters.  

<img width="280" alt="New filters " src="https://user-images.githubusercontent.com/92091904/165285894-b7101665-dd40-4f44-8087-e9e293fabc95.png">

- The summary line was added in from the Accordion component to shows how many items the user has selected in which was originally underneath 'Filter by'. 

<img width="273" alt="New filters expanded " src="https://user-images.githubusercontent.com/92091904/165285933-4e1011f9-baa7-400d-9861-ce070aa1b6a3.png">

**Comment by maddyow** [comment]
We've used basic filters on the [Campaigns hub](https://campaignresources.dhsc.gov.uk/campaigns/) to narrow down a page of campaign results. 
![CRC v3 - Campaigns hub filters](https://user-images.githubusercontent.com/44677400/171204608-bf516e8a-5c53-45bf-8e6c-06a84dc98f7d.png)

In recent research sessions, the filters have been interacted with and useful to some – particularly with users who are familiar with the content (this is more of a B2B site). There were some expectations for things that weren’t on the site, for example specific health conditions so the labels and categories are very important (as well as the drop down options). There is also a more advanced search on this site so filters are used to help search at a topic only level.

Also noted in https://github.com/nhsuk/nhsuk-service-manual-community-backlog/issues/217

**Comment by CharlotteDowns** [comment]
We've just [added some guidance on how to 'share findings about your users' with us](https://design-system.service.gov.uk/community/share-research-findings/) 📝. This will help us learn more about how your users filter a list within your service.
